### PR TITLE
Beginning of ROCKET implementation

### PR DESCRIPTION
**Pull request by ABostrom**
implemented random kernel construction

using this as the python reference: https://github.com/angus924/rocket/blob/master/code/rocket_functions.py

#434

### CODE REVIEW

**Comment by MatthewMiddlehurst** [comment]
Added multivariate functionality similar to the sktime version and multithreading for the transformer. 

Implemented a classifier using the ROCKET transformer with basic contracting (only the transformer, not the classifier).

Compared to the sktime transformer and found it was not significantly different using a 1-NN classifier over 30 folds.

**Comment by ABostrom** [comment]
This looks pretty good Matt. Good job on wrapping it up, and making a proper classifier around the transform.

**Comment by James-Large** [comment]
This looks all good to me, couple points: 

1) in the rocket transformer, could the single threaded fit be easily implemented in terms of a FitThread.call() too? Obviously just in the current thread though as opposed to via the executor service 
2) what's the current state of the reproducibility to published results? If not sig similar on a reasonable number of datasets etc, and especially because randf is being taken as the default classifier, a note of this should probs be made in the class jdoc. Default assumption would be that default constructor of RocketClassifier constructs something that is true to the published results, should state if otherwise

**Comment by MatthewMiddlehurst** [comment]
Regarding threading we could get rid of one or the other to clean up/save space tbh. I realised mid-way that fit probably takes so little time that its not really worth thinking about unless you are creating a ridiculous amount of kernels.

RandomForest was not meant to be the default, my bad. I was using it for testing as weka Logistic takes an insane amount of memory. I have replaced it with a wrapper for liblinear.

In regards to comparing to published/sktime results, not really close at the moment so I wouldn't recommend merging. I saved the tsml ROCKET transform to file and used the sklearn classifier with no significant difference, so again its not that. The sklearn classifier we are comparing against is RidgeClassifierCV (https://github.com/scikit-learn/scikit-learn/blob/0fb307bf3/sklearn/linear_model/_ridge.py).

**Comment by ABostrom** [comment]
@MatthewMiddlehurst so If i understand you correctly. The Transform is 100% accurate to published results.
Its the RidgeClassifier we can't replicate?

**Comment by MatthewMiddlehurst** [comment]
Yes. 

These are the results for the authors sktime implementation (ROCKET_py), our implementation using the current LibLinearClassifer setup (ROCKET) and the python ROCKET transform with a default sklearn LogisticRegressionCV classifier (ROCKET_Lpy). 87 datasets 10 folds.


. | ROCKET_py | ROCKET | ROCKET_Lpy
-- | -- | -- | --
Accuracy | 0.88517 | 0.856162 | 0.848698
Acc Rank | 1.247126 | 2.149425 | 2.603448

The difference is significant.

**Comment by MatthewMiddlehurst** [comment]
I bit the bullet and just implemented a cut down version of RidgeClassiferCV. Reproduced the results as well as we are ever going to get in my opinion.

Im using Nd4j for matrix maths, which is a very large dependency. If anyone has any alternative suggestions that would be good (or how to configure it so my JAR isent an extra 140-380MB in size).

**Comment by ABostrom** [comment]
This is actually really useful. I've been discussing with Tony about using Nd4j in TSML for sometime now. 
In my opinion this proves how useful it is. Well done.

**Comment by James-Large** [comment]
Looks solid to me, happy to merge. nd4j will make some things a lot easier for sure, despite it's size. Missed meeting earlier for doc appointment buts sounds like the inclusion has been given the go ahead generally

**Comment by MatthewMiddlehurst** [comment]
Yeah. From my POV this is basically good to go apart from some documentation/comments.

Since we are adding dependencies, I’m going to add libsvm before we merge. There seems to be some duplicates in both the lib folder and gradle file which I’m not sure are necessary also. A good chance to make dependency changes if anyone was hesitant before.

**Comment by MatthewMiddlehurst** [comment]
Also worth noting we are currently merging into dev, I think we decided to discontinue that at some point in favour of master?
